### DISCUSSION TITLE
Optimizing shell startup time

### DISCUSSION THREAD

**Discussion by twpayne** [twpayne/chezmoi#2971] [Show and tell]
@mono0x designed [a really neat trick to speed up your startup scripts](https://blog.mono0x.net/2023/04/29/optimize-rcfiles-using-chezmoi/).

The post is in Japanese but you can read it with Google Translate.

**Comment by mono0x** [comment]
Thank you for developing such powerful software and considering my idea.

The idea is to cache scripts generated by specific commands.

For instance, Homebrew's `brew shellenv` command generates a script that shell rc files should load. This result can be cached using chezmoi's template feature.

```
{{ output "/opt/homebrew/bin/brew" "shellenv" | trim }}
```

This is equivalent to `eval "$(/opt/homebrew/bin/brew shellenv)"` as a result. However, the `brew shellenv` command is executed during the `chezmoi apply` phase. This eliminates the overhead of executing the `brew shellenv` command when starting the shell.

Moreover, the cached script can be easily refreshed by running `chezmoi apply`.

Thanks to chezmoi, we have introduced a handy and low-cost cache mechanism.

**Reply by twpayne** [reply]
Another nice side-effect of this solution is that `chezmoi diff` will show you what has changed, which you don't see if you use `eval $(/opt/homebrew/bin/brew shellenv)`.

**Reply by Lockszmith-GH** [reply]
Interesting, does the template get evaluated everytime `chezmoi apply` is run? if so, then yes - this is a great solution.

**Reply by mono0x** [reply]
> does the template get evaluated everytime chezmoi apply is run?

Yes. `output` function executes the command every time. I think that such behavior helps to keep the cache up-to-date.

https://www.chezmoi.io/reference/templates/functions/output/#:~:text=The%20execution%20occurs%20every%20time%20that%20the%20template%20is%20executed.

**Reply by halostatue** [reply]
There may be some drawbacks depending on the tool. I’m specifically thinking of `brew shellenv`, which uses `/bin/ps` to select the currently running shell and cannot be overridden. If you generate shell caches for multiple types of shells, you will get unanticipated results.

I’ve opened Homebrew/brew#15358 as a suggestion to make this work for this (admittedly *narrow*) use case.

What’s nice is that you can write your shell scripts to be smart, something like:

```bash
if [[ -s "${HOME}/.cache/shell-cache/brew.sh" ]]; then # exists and non-empty
  source "${HOME}/.cache/shell-cache/brew.sh"
else
  command -v brew >/dev/null 2>/dev/null && eval "$(brew shellenv)"
fi
```

It would be necessary to use `-s` if you have conditional caches (e.g., caching `starship shell zsh` should only run if you have `starship` installed).

**Reply by mono0x** [reply]
To run `brew shellenv` in the intended shell, it seems that the following workaround can be used.

```
{{ output "fish" "-c" "/opt/homebrew/bin/brew shellenv" | trim }}
```

> It would be necessary to use -s if you have conditional caches (e.g., caching starship shell zsh should only run if you have starship installed).

Yes, but you can also check if the app is installed using `lookPath` function.

```
{{- if lookPath "starship" -}}
{{ output "starship" "init" "zsh" "--print-full-init" | trim }}
{{- end -}}
```

**Reply by halostatue** [reply]
That does seem to work. I’m still hopeful that they will allow the extension to shellenv since I may want to build something that supports powershell *without* having to have powershell installed. (Yes, I know I could probably do some script shenanigans to get and transform the values as required, but…)

**Reply by halostatue** [reply]
I’ve opened Homebrew/brew#15363 to add this feature.
